lib/cppapi/include/atscppapi/Continuation.h

```diff
@@ -0,0 +1,142 @@
+/**
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+/**
+ * @file Continuation.h
+ * @brief Wrapper class for TS API type TSCont.
+ */
+
+#pragma once
+
+#include <ts/ts.h>
+
+namespace atscppapi
+{
+class Continuation
+{
+public:
+  using Mutex = TSMutex;
+
+  using Action = TSAction;
+
+  // Create continuation, mutexp may be nullptr.
+  //
+  Continuation(Mutex mutexp) : _cont(TSContCreate(_generalEventFunc, mutexp)) { TSContDataSet(_cont, static_cast<void *>(this)); }
+
+  // Create "empty" continuation, can only be populated by move assignement.
+  //
+  Continuation() : _cont(nullptr) {}
+
+  TSCont
+  asTSCont() const
+  {
+    return _cont;
+  }
+
+  // Get mutex (for "non-empty" continuation).
+  //
+  Mutex
+  mutex()
+  {
+    return _cont ? TSContMutexGet(_cont) : nullptr;
+  }
+
+  void
+  destroy()
+  {
+    if (_cont) {
+      TSContDestroy(_cont);
+      _cont = nullptr;
+    }
+  }
+
+  virtual ~Continuation()
+  {
+    if (_cont) {
+      TSContDestroy(_cont);
+    }
+  }
+
+  // No copying.
+  Continuation(const Continuation &) = delete;
+  Continuation &operator=(const Continuation &) = delete;
+
+  // Moving allowed.
+  Continuation(Continuation &&that)
```

Comment:
I think you should `std::swap` the elements, to make sure `that` is cleared and doesn't (for instance) double delete the `TSCont` when it goes out of scope.

Reply:
Sorry I'm drawing a complete blank here, no idea what the point would be.  I do clear that by setting thtat._cont to the nullptr .